# PYTHON – WORKSHEET 12(DATA CLEANING)
## ANSWER KEYS

In [1]:
import pandas as pd
df=pd.read_csv(r"C:\Users\devanshu suri\Downloads\PYTHON-WORKSHEET-12\data.csv")
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St    2.0    h        NaN     SS   
1        Abbotsford        85 Turner St    2.0    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St    3.0    u        NaN     VB   
4        Abbotsford        5 Charles St    3.0    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23673  Wyndham Vale     25 Clitheroe Dr    3.0    u        NaN     PN   
23674  Wyndham Vale    19 Dalrymple Bvd    4.0    h        NaN      S   
23675     Yallambie        17 Amaroo Wy    4.0    h  1100000.0      S   
23676    Yarraville          6 Agnes St    4.0    h  1285000.0     SP   
23677    Yarraville       33 Freeman St    4.0    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
23673      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23674  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23675     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23676        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23677        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0     1900.0        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0     1900.0        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
23673     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23674       NaN           NaN        NaN          NaN -37.87882   144.60184   
23675       NaN           NaN        NaN          NaN -37.72006   145.10547   
23676     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23677       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
23673   Western Metropolitan        5262.0  
23674   Western Metropolitan        5262.0  
23675  Northern Metropolitan        1369.0  
23676   Western Metropolitan        6543.0  
23677   Western Metropolitan        6543.0  

[23678 rows x 21 columns]

 **1. Determine the no. of rows and no. of columns in the dataset:**

In [2]:
shape=df.shape
print("\nTotal Number of Rows : ",shape[0])
print("Total Number of columns : ", shape[1])


Total Number of Rows :  23678
Total Number of columns :  21


 **2. Display the datatypes of each column**

In [3]:
print("\n\nData Types of each column :\n", df.dtypes)



Data Types of each column :
 Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object


 **3. Determine the row indexes of rows having all null values (i.e. empty rows).**

In [4]:
print("\nthe row indexes of rows having all null values:\n", df[df.isna().all(axis=1)].index,"\n\n")




the row indexes of rows having all null values:
 Int64Index([   31,  1145,  1348,  2119,  3277,  3288,  3745,  3752,  3798,
             3855,
            ...
            21220, 21343, 21602, 22003, 22109, 22288, 22704, 22918, 23153,
            23341],
           dtype='int64', length=131) 




 **4. Determine the total number of empty rows and their percentage in whole dataset and remove them.**

In [5]:
empty_rows_len=df.isna().all(axis=1).sum()
total_rows_length=df.shape[0]
empty_rows_percentage=(empty_rows_len / total_rows_length)*100

print("Total Number of empty rows: ",empty_rows_len)

print("Empty rows Percentage: ", empty_rows_percentage)

Total Number of empty rows:  131
Empty rows Percentage:  0.5532561871779711


In [6]:
df.drop(df[df.isna().all(axis=1)].index,inplace=True)
df.shape

(23547, 21)

**5. Determine the column names having missing values.**


In [7]:
df.columns[df.isnull().any()]


Index(['Price', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

**6. Determine the missing values and their percentage in each column..**


In [8]:
print("\nTotal number of empty rows in each column:\n", df.isnull().sum(),"\n\n")

percent_missing = round((df.isnull().sum()*100)/len(df),2)

print("\nTotal number of empty rows percentage in each column:\n", percent_missing,"\n\n")




Total number of empty rows in each column:
 Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64 



Total number of empty rows percentage in each column:
 Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude

**7. Determine the row indexes having missing values.**

In [9]:
row_ind=[]
for index,row in df.iterrows():
    row_na=row.isnull()
    if row_na.any():
        row_ind.append(index)
        
print(row_ind)

[0, 1, 3, 5, 7, 8, 9, 10, 12, 13, 15, 16, 17, 19, 20, 21, 22, 23, 26, 27, 28, 29, 32, 34, 35, 37, 40, 41, 42, 46, 47, 48, 49, 51, 53, 54, 55, 56, 61, 63, 65, 66, 70, 75, 76, 77, 78, 79, 80, 81, 85, 88, 90, 92, 93, 94, 95, 96, 97, 98, 100, 102, 103, 106, 108, 111, 112, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 131, 132, 133, 135, 138, 140, 141, 142, 144, 145, 147, 149, 151, 152, 154, 155, 157, 158, 159, 161, 164, 165, 166, 168, 170, 171, 172, 174, 175, 176, 180, 181, 182, 183, 185, 186, 188, 189, 190, 191, 193, 195, 197, 198, 199, 200, 201, 204, 207, 208, 209, 210, 211, 212, 213, 214, 219, 221, 222, 223, 224, 225, 226, 228, 229, 235, 238, 240, 241, 244, 246, 248, 249, 251, 253, 254, 255, 257, 258, 259, 261, 264, 266, 267, 268, 269, 271, 272, 273, 274, 275, 276, 278, 280, 281, 282, 284, 287, 288, 289, 291, 292, 294, 295, 297, 298, 299, 301, 303, 304, 305, 306, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 324, 325, 326, 327, 328, 329, 3

**8. Determine the missing values and their percentage in each row.**

In [10]:
print("\n missing values in each row:\n", df.isnull().sum(axis=1),"\n\n")

percent_missing = round((df.isnull().sum(axis=1)*100)/len(df.columns),2)

print("\nTotal number of empty rows percentage in each column:\n", percent_missing,"\n\n")




 missing values in each row:
 0        3
1        2
2        0
3        3
4        0
        ..
23673    2
23674    8
23675    4
23676    1
23677    2
Length: 23547, dtype: int64 



Total number of empty rows percentage in each column:
 0        14.29
1         9.52
2         0.00
3        14.29
4         0.00
         ...  
23673     9.52
23674    38.10
23675    19.05
23676     4.76
23677     9.52
Length: 23547, dtype: float64 




**9. Determine the column names having more than 30% missing values and remove them from the dataset.**

In [11]:
pc_null = df.isnull().sum() / len(df)
missing_features = pc_null[pc_null > 0.30].index
print("\nThe column  having more than 30% missing values are \n:",missing_features)
df.drop(missing_features, axis=1, inplace=True)
print("\nAfter remove more than 30% missing value columns remaining columns left are \n:",df.columns)
df.shape


The column  having more than 30% missing values are 
: Index(['BuildingArea', 'YearBuilt', 'CouncilArea'], dtype='object')

After remove more than 30% missing value columns remaining columns left are 
: Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'Lattitude', 'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')


(23547, 18)

**10. Determine the number and percentage of rows having more than 5 missing values. If the percentage of such
rows is less than 20% then remove all these rows**

In [12]:
d = df.loc[df.isnull().sum(axis=1)>5]
a=len(d)
b=round(len(d)/df.shape[0] * 100,2)
print(f'number and percentage of rows having more than 5 missing values are {a} and  {b}%')

number and percentage of rows having more than 5 missing values are 4278 and  18.17%


In [13]:
# percentage of such rows is less than 20% so remove all these rows
df.drop(df.loc[df.isnull().sum(axis=1)>5].index,axis=0,inplace=True)
df.shape

(19269, 18)

**11. Display the description of data and percentage of missing values still left column wise.**

In [14]:
df.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  19269.000000  1.508600e+04  19269.000000  19269.000000  19066.000000   
mean       2.973117  1.066695e+06     10.204365   3110.015777      2.951956   
std        0.978969  6.389331e+05      6.065786     93.730487      0.996032   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.400000e+05      6.100000   3049.000000      2.000000   
50%        3.000000  9.000000e+05      9.200000   3102.000000      3.000000   
75%        4.000000  1.320000e+06     13.000000   3150.000000      4.000000   
max       12.000000  9.000000e+06     48.100000   3977.000000     30.000000   

           Bathroom           Car       Landsize     Lattitude    Longtitude  \
count  19063.000000  18921.000000   17410.000000  19243.000000  19243.000000   
mean       1.570897      1.626235     551.783458    -37.812434    145.000287   
std        0.712684      0.974048    3544.288014      0.079926      0.106071   
min        0.000000      0.000000       0.000000    -38.184180    144.431620   
25%        1.000000      1.000000     181.000000    -37.859300    144.939300   
50%        1.000000      2.000000     448.000000    -37.809700    145.004300   
75%        2.000000      2.000000     656.000000    -37.759800    145.063100   
max       12.000000     26.000000  433014.000000    -37.407580    145.526350   

       Propertycount  
count   19269.000000  
mean     7490.249520  
std      4319.566673  
min       129.000000  
25%      4442.000000  
50%      6567.000000  
75%     10331.000000  
max     21650.000000

In [15]:
percent_missing = round((df.isnull().sum()*100)/len(df),2)

print("\nTotal number of empty rows percentage in each column:\n", percent_missing,"\n\n")




Total number of empty rows percentage in each column:
 Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64 




**12. Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values?
Give a brief explanation and treat the missing values in 'Price' column**

We have treat the missings values in price column with mean() of its values because if we drop all nan values in price column we lost large amount of information so we have three opton to replace nan values with mean,median and mode as distribution for the price column data is symmetric in nature so i'll use mean().

In [16]:
#replacing missing values with mean in price
df['Price'].fillna(df['Price'].mean(), inplace=True)


**13. Now check the percentage of missing values in each column again, if missing values exist, treat it with the most
suitable method.**

In [17]:
percent_missing = round((df.isnull().sum()*100)/len(df),2)

print("\nTotal number of empty rows percentage in each column:\n", percent_missing,"\n\n")


Total number of empty rows percentage in each column:
 Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.81
Landsize         9.65
Lattitude        0.13
Longtitude       0.13
Regionname       0.00
Propertycount    0.00
dtype: float64 




In [18]:
#Longtitude and Lattitude contains very less nan values so we can drop these nan values
df = df[df['Longtitude'].notna()]
df = df[df['Lattitude'].notna()]
#replacing missing values with mean in Landsize
df['Landsize'].fillna(df['Landsize'].mean(), inplace=True)
#replacing missing values with mode in Car
df['Car'].fillna(df['Car'].mode()[0], inplace=True)
#replacing missing values with mode in Bedroom2
df['Bedroom2'].fillna(df['Bedroom2'].mode()[0], inplace=True)
#replacing missing values with mode in Bathroom
df['Bathroom'].fillna(df['Bathroom'].mode()[0], inplace=True)


In [19]:
# Lets check again percentage of missing values after treating nan values in each field
percent_missing = round((df.isnull().sum()*100)/len(df),2)

print("\nTotal number of empty rows percentage in each column:\n", percent_missing,"\n\n")


Total number of empty rows percentage in each column:
 Suburb           0.0
Address          0.0
Rooms            0.0
Type             0.0
Price            0.0
Method           0.0
SellerG          0.0
Date             0.0
Distance         0.0
Postcode         0.0
Bedroom2         0.0
Bathroom         0.0
Car              0.0
Landsize         0.0
Lattitude        0.0
Longtitude       0.0
Regionname       0.0
Propertycount    0.0
dtype: float64 




**14. Now since you have completed your data cleaning, determine the number and percentage of columns removed.**

* We have removed 3 columns which had more than 30% missing values  

In [20]:
print("Number of Columns removed:",len(missing_features))
print("Percentage of the columns removed",(len(missing_features)/len(df.columns))*100)

Number of Columns removed: 3
Percentage of the columns removed 16.666666666666664


**15. Determine the number and percentage of rows removed in whole process. Also comment briefly that according
to you the amount of data lost is justified or not.**

In [21]:
df.shape

(19243, 18)

In [22]:
df1=pd.read_csv(r"C:\Users\devanshu suri\Downloads\PYTHON-WORKSHEET-12\data.csv")
df1.shape

(23678, 21)

In [23]:
number=df1.shape[0]-df.shape[0]
percentage=round(number/df1.shape[0] * 100,2)
print(f'number and percentage of rows removed in whole process are {number} and  {percentage}%')


number and percentage of rows removed in whole process are 4435 and  18.73%


* Percentage of rows removed in whole process are 18.73 % which is significally loss of data which is not good for modelling.

# End of Document